File: src/model/codec.ts
import type { Model, ModelClass } from "./model";

export interface ModelCodec {
  parseOne<T extends Model>(cls: ModelClass<T>, payload: unknown): T | null;
  parseMany<T extends Model>(cls: ModelClass<T>, payload: unknown): T[];
  serialize?(model: Model, options: { dirtyOnly: boolean }): unknown;
}

export function codecFor<T extends Model>(cls: ModelClass<T>): ModelCodec {
  return cls.codec ?? jsonApiCodec();
}

function jsonApiCodec(): ModelCodec {
  return requireJsonApiCodec();
}

let cachedJsonApiCodec: ModelCodec | null = null;

function requireJsonApiCodec(): ModelCodec {
  if (cachedJsonApiCodec) return cachedJsonApiCodec;
  throw new Error("jsonApiCodec has not been installed");
}

export function installDefaultCodec(codec: ModelCodec): void {
  cachedJsonApiCodec = codec;
}

File: opal/lib/swill/model/codec.rb
# frozen_string_literal: true

module Swill
  module Model
    module Codec
      def parse_one(_model_class, _payload)
        raise NotImplementedError
      end

      def parse_many(_model_class, _payload)
        raise NotImplementedError
      end

      def serialize(_model, dirty_only: false)
        raise NotImplementedError
      end
    end

  end
end
